datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -603,6 +604,73 @@ impl LogicalPlan {
         Ok(new_plan)
     }
 
+    /// Walk the logical plan, find any `PlaceHolder` tokens, and return a map of their IDs and DataTypes
+    pub fn get_parameter_types(
+        &self,
+    ) -> Result<HashMap<String, Option<DataType>>, DataFusionError> {
+        struct ParamTypeVisitor {
+            param_types: HashMap<String, Option<DataType>>,
+        }
+
+        struct ExprParamTypeVisitor {
+            param_types: HashMap<String, Option<DataType>>,
+        }
+
+        impl ExpressionVisitor for ExprParamTypeVisitor {
+            fn pre_visit(
+                mut self,
+                expr: &Expr,
+            ) -> datafusion_common::Result<Recursion<Self>>
+            where
+                Self: ExpressionVisitor,
+            {
+                if let Expr::Placeholder { id, data_type } = expr {
+                    let _ = self.param_types.insert(id.clone(), data_type.clone());
```

Comment:
Should we error  here if the parameter was previously referenced with a different datatype 🤔 (aka don't ignore `_`?)

Reply:
I think it's valid to have multiple `$1`s in a single statement. Maybe we should check the type and error if it differs?